orangecontrib/text/pubmed.py

```diff
@@ -0,0 +1,480 @@
+import os
+import re
+import shelve
+import warnings
+import itertools
+from datetime import datetime
+
+import numpy as np
+from Bio import Entrez
+from Bio import Medline
+from validate_email import validate_email
+
+from Orange.canvas.utils import environ
+from Orange.data import StringVariable, DiscreteVariable, Domain
+from orangecontrib.text.corpus import Corpus
+
+PUBMED_TEXT_FIELDS = [
+    ('authors', 'FAU'),
+    ('article_title', 'TI'),
+    ('mesh_headings', 'MH'),
+    ('abstract', 'AB'),
+    ('pub_date', 'DP'),
+]
+
+
+def _join_if_list(value):
+    if isinstance(value, list):
+        return ' '.join(value)
+    return value
+
+
+def _mesh_headings_to_class(mesh_headings):
+    """Extract a class from the mesh headings.
+
+    We take the first mesh heading and extract it to use as a class.
+
+    Args:
+        mesh_headings (str): The string containing all the mesh headings.
+
+    Returns:
+        str: The class value.
+    """
+    # e.g. heading1/heading2,heading3/*heading4
+    try:
+        regex = re.compile(r'^(.\w*?)\W')
+        class_mesh_groups = regex.search(mesh_headings[0])
+        return class_mesh_groups.groups()[0]
+    except:
+        return None
+
+
+def _date_to_iso(date):
+    possible_date_formats = [
+        '%Y %b %d',
+        '%Y %b',
+        '%Y',
+    ]
+
+    season_mapping = {
+        'fall': 'Sep',
+        'winter': 'Dec',
+        'spring': 'Mar',
+        'summer': 'Jun',
+    }
+
+    for date_format in possible_date_formats:
+        try:
+            return datetime.strptime(date, date_format).date().isoformat()
+        except Exception:
+            continue  # Try the next format.
+
+    date = date.lower()
+    # Seasons to their respective months.
+    for season, month in season_mapping.items():
+        date = date.replace(season, month)
+    try:  # Try month range (e.g. Jan-Mar).
+        date = date.split('-')[0]  # 2015 Sep-Dec --> 2015 Sep
+        return datetime.strptime(
+                date,
+                possible_date_formats[1]
+        ).date().isoformat()
```

Comment:
changing date from datetime to iso, str etc. is being used here, in scraper widget  as well as NYT
instead of writing the functions again and again we can have a file with most of the commonly used functions